Add pagination within the BigQuery driver

robdaemon opened the pull request:
The BigQuery API has a specific method of pagination using a "page
token" in a query result. Use this to retrieve the entire result set in
a lazy fashion.

Resolves #6960

[ci bigquery]

Commit:
Add pagination within the BigQuery driver

The BigQuery API has a specific method of pagination using a "page
token" in a query result. Use this to retrieve the entire result set in
a lazy fashion.

Resolves #6960

[ci bigquery]

sbelak commented on modules/drivers/bigquery/src/metabase/driver/bigquery.clj:
```diff
@@ -200,35 +225,30 @@
+                                                                                    next-page-token)]
+                                (fetch-page next-resp)))))]
+        (for [^TableRow row (fetch-page response)]
+          (for [[^TableCell cell, parser] (partition 2 (interleave (.getF row) parsers))]
```
Wouldn't `(map (fn [cell parser] ...) (.getF row) parsers)` be clearer here?

robdaemon replied:
Maybe? The nested `for` loops were already present before my change - I added the `fetch-page` stuff.

camsaul replied:
I did the nested `for` because I thought *that* was clearer but I can see how `map` would be clearer... I could go either way if you want to change it

Commit:
Fixed tests to work with new interface

Added ability to use a service account json for BigQuery testing in
development

Still need to add a test specifically for pagination

[ci bigquery]

Commit:
Testing that pagination occurs

Due to lack of metrics on calls made, a dynamic callback was added so
that call counts can be tracked to indicate that pagination had
occurred.

Also, introduces a maximum row count to each BigQuery call. Leaving at
an arbitrarily high number - 20,000. BigQuery will automatically stop at
a max of 10 MB per page, regardless of this number, but the setting is
used in tests to ensure what pagination is doing.

[ci bigquery]

camsaul commented on modules/drivers/bigquery/src/metabase/driver/bigquery.clj:
```diff
@@ -144,10 +144,18 @@
 
 (def ^:private ^:const ^Integer query-timeout-seconds 60)
 
+(def ^:private ^:dynamic ^Long max-results
+  "Maximum number of rows to return per request in a query."
+  20000)
```
we have an `absolute-max-results` constant defined somewhere, does it make sense to use that instead of having a different max for BigQuery?

Also, I think 20k is a little low... some people might want to download a huge dataset

robdaemon replied:
this is per page, not total - let me rename it for clarity to `max-results-per-page`

flamber replied:
Is the new pagination getting past the BQ API limitation of 10MB? Or would it be better to apply a smaller page max?
https://cloud.google.com/bigquery/quotas#jobsquery_requests

robdaemon replied:
So, `max-results-per-page` doesn't overrule the 10MB limit. Pages can be a maximum of 10MB, or `max-results-per-page`, whichever threshold is met first.

On the HN dataset that Damon has, I would get approximately 22,000 rows per page.

With this change (pagination), I managed to download a CSV approximately 600 mb in size before the dev-environment timeout of 30 seconds hit.

In BigQuery, the "best" way to do a query is to submit the query as a job, and then subsequently call the GetQueryResults API repeatedly to paginate through the results with that job ID and project ID.

flamber replied:
Okay, OP on #6960 was only getting 1,759 rows, so that was just why I wanted to ask.

camsaul commented on modules/drivers/bigquery/src/metabase/driver/bigquery.clj:
```diff
@@ -163,6 +171,8 @@
       :else
       (throw (ex-info "Query timed out." (into {} response))))))
 
+;;
+;;
```
leftovers?

camsaul commented on modules/drivers/bigquery/src/metabase/driver/bigquery.clj:
```diff
@@ -200,35 +237,30 @@
-       (for [^TableRow row (.getRows response)]
-         (for [[^TableCell cell, parser] (partition 2 (interleave (.getF row) parsers))]
-           (when-let [v (.getV cell)]
+       (letfn [(fetch-page [^GetQueryResultsResponse response]
+                           (lazy-cat
+                            (.getRows response)
+                            (when-let [next-page-token (.getPageToken response)]
+                              (with-finished-response [next-resp (get-query-results (database->client database)
+                                                                                    (.getProjectId (.getJobReference response))
+                                                                                    (.getJobId (.getJobReference response))
+                                                                                    next-page-token)]
```
This should be indented like 

```clj
(letfn [(fetch-page [...]
          (lazy-cat ...)
```

camsaul reviewed (approved):
Other than the `max-results` which I think we should carefully consider this LGTM

Commit:
Clarify max-results as being per page

indentation and other small review feedback

[ci bigquery]
